### PR TITLE
ENH: Energy request waiter

### PR DESCRIPTION
**Pull request by ZLLentz**
## Description

There's a mode that the energy request PVs can run in where done/not-done information is broadcast by ACR to one of a number of status PVs. This PR adds a mode to `BeamEnergyRequest` where motion will wait and report done based on that given PV instead of using the "done immediately" mechanisms.

This is accomplished by using (abusing?) the `__new__` mechanics to pick which subclass to substitute for `BeamEnergyRequest` based on the input arguments. These subclasses each implement `done` in a different way. The parent `BeamEnergyRequest` class continues to implement shared behavior e.g. pv names of the setpoint PV. This is similar to our handling of `PytmcSignal`. At the cost of some implementation complexity, compared to making a factory functions this allows all variants to be isinstance checked against the "factory".

This is fully backwards-compatible with the previous implementation.

## Motivation and Context

Request from James Cryan and it just seems like a better way to handle these overall
https://slac.slack.com/archives/C9RP2PQHM/p1652145320558209
https://jira.slac.stanford.edu/browse/LCLSPC-402

## How Has This Been Tested?

Interactively I verified that the PVs looked correct.
The fake devices seem to pass the tests I've set up for them.
This probably needs some test in TMO

## Where Has This Been Documented?

Added pre-release notes

## Pre-merge checklist
- [x] Code works interactively
- [x] Code contains descriptive docstrings, including context and API
- [x] New/changed functions and methods are covered in the test suite where possible
- [x] Test suite passes locally
- [x] Test suite passes on travis
- [x] Ran docs/pre-release-notes.sh and created a pre-release documentation page
- [x] Pre-release docs include context, functional descriptions, and contributors as appropriate

### CODE REVIEW

**Comment by ZLLentz** [comment]
Seems like I did break something in CCM- sigh

**Review by klauer — APPROVED** [review]
This is an interesting one - with unfortunate PV functionality...

I think you did the best you could with the cards that were dealt. Nothing jumped out as problematic or terribly odd, and everything looked well-documented and reasonably well-tested.

**Review by tangkong — APPROVED** [review]
Looks good to me.  Strange to have to do this, but it looks solid.

File: docs/source/upcoming_release_notes/1013-energy-request-waiter.rst [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+--------
+- N/A
+
+Maintenance
+-----------
+- N/A
+
+Contributors
+------------
+- N/A
```
[the remark this answers is not in the document]
Comment (ZLLentz) [reply docs/source/upcoming_release_notes/1013-energy-request-waiter.rst:33]: lol oops

**Comment by ZLLentz** [comment]
> Strange to have to do this

I agree, this all stems from having multiple flexible configurations for running the same device.

**Pull request merged**